pkg/sources/travisci/travisci.go

```diff
@@ -0,0 +1,222 @@
+	jobPool  *errgroup.Group
+	sources.Progress
+	client *travis.Client
+	sources.CommonSourceUnitUnmarshaller
+	returnAfterFirstChunk bool
+}
+
+// Ensure the Source satisfies the interfaces at compile time.
+var _ sources.Source = (*Source)(nil)
+var _ sources.SourceUnitUnmarshaller = (*Source)(nil)
+
+// Type returns the type of source.
+// It is used for matching source types in configuration and job input.
+func (s *Source) Type() sourcespb.SourceType {
+	return SourceType
+}
+
+func (s *Source) SourceID() sources.SourceID {
+	return s.sourceId
+}
+
+func (s *Source) JobID() sources.JobID {
+	return s.jobId
+}
+
+// Init returns an initialized TravisCI source.
+func (s *Source) Init(ctx context.Context, name string, jobId sources.JobID, sourceId sources.SourceID, verify bool, connection *anypb.Any, concurrency int) error {
+	s.name = name
+	s.sourceId = sourceId
+	s.jobId = jobId
+	s.verify = verify
+	s.jobPool = &errgroup.Group{}
+	s.jobPool.SetLimit(concurrency)
+
+	var conn sourcespb.TravisCI
+	if err := anypb.UnmarshalTo(connection, &conn, proto.UnmarshalOptions{}); err != nil {
+		return errors.WrapPrefix(err, "error unmarshalling connection", 0)
+	}
+
+	switch conn.Credential.(type) {
+	case *sourcespb.TravisCI_Token:
+		if conn.GetToken() == "" {
+			return errors.New("token is empty")
+		}
+		s.client = travis.NewClient(baseURL, conn.GetToken())
+		s.client.HTTPClient = common.RetryableHttpClientTimeout(3)
+
+		user, _, err := s.client.User.Current(ctx, nil)
+		if err != nil {
+			return errors.WrapPrefix(err, "error getting testing travis client", 0)
+		}
+		ctx.Logger().V(2).Info("authenticated to Travis CI with user", "username", user.Login)
+	}
+
+	return nil
+}
+
+func (s *Source) Enumerate(ctx context.Context, reporter sources.UnitReporter) error {
+	for repoPage := 0; ; repoPage++ {
+		repositories, _, err := s.client.Repositories.List(ctx, &travis.RepositoriesOption{
+			Limit:  pageSize,
+			Offset: repoPage * pageSize,
+		})
+		if err != nil {
+			if repoPage == 0 {
+				return fmt.Errorf("error listing repositories: %w", err)
+			} else {
+				err = reporter.UnitErr(ctx, err)
+				if err != nil {
+					return fmt.Errorf("error reporting error: %w", err)
+				}
+			}
+		}
+
+		if len(repositories) == 0 {
+			break
+		}
+
+		for _, repo := range repositories {
+			err = reporter.UnitOk(ctx, sources.CommonSourceUnit{
+				ID: strconv.Itoa(int(*repo.Id)),
+			})
+			if err != nil {
+				return fmt.Errorf("error reporting unit: %w", err)
+			}
+			ctx.Logger().V(2).Info("enumerated repository", "id", repo.Id, "name", repo.Name)
+		}
+	}
+
+	return nil
+}
+
+// ChunkUnit implements SourceUnitChunker interface.
+func (s *Source) ChunkUnit(ctx context.Context, unit sources.SourceUnit, reporter sources.ChunkReporter) error {
+	repo, _, err := s.client.Repositories.Find(ctx, unit.SourceUnitID(), nil)
+	if err != nil {
+		return fmt.Errorf("error finding repository: %w", err)
+	}
+	logger := ctx.Logger().WithValues("repo", *repo.Name)
```

Comment:
No, it depends on the implementation but that's generally the idiom.

Re: logger, I'm not sure but should to caller populate those (source id etc) consistently rather than the source being responsible?

Reply:
Tested the logger locally and this is what's printed, so it already has context (though missing `source_id` it seems).

```json
{
  "level": "info-2",
  "ts": "2023-10-18T11:54:13-07:00",
  "logger": "trufflehog",
  "msg": "scanning repository",
  "job_id": 1,
  "source_manager_worker_id": "G26hU",
  "source_type": "SOURCE_TYPE_TRAVISCI",
  "source_name": "trufflehog - Travis CI",
  "unit": "26155351",
  "repo": "test-repo"
}
```

Reply:
> No, it depends on the implementation but that's generally the idiom.

If we're not 100% confident that `repo` won't be `nil` (maybe `Find` is misbehaving, I guess) do you think it's worth adding a `nil` check for `repo` to avoid a potential panic?